### PR TITLE
Crowdin integration

### PR DESCRIPTION
**Pull request by Nadahar**
@UniversalMediaServer/developers 
The POM is in a testing state here now, but I'm getting closer to finish the crowdin integration. I've had to fork [this Maven plugin] (https://github.com/glandais/crowdin-maven) and done some rather major modifications which you can see [here] (https://github.com/Nadahar/crowdin-maven).

Let me just start by saying: Aaaaaaargh :boom::facepunch::sweat_drops::punch::goberserk::skull::rage::collision: doesn't even begin to describe how much I hate Maven:exclamation: It's one undocumented piece of a complete mess.

But, with some hacking I've been able to get things to work more or less as planned. In the plugin I've this far implemented these goals:
* ```push``` - upload ```messages.properties``` to crowdin
* ```build``` - build or "compile" the translations database into downloadable files
* ```fetch``` - download the language files from crowdin into a temporary folder

What I have left is:
* ```apply``` - transfer (and rename, modify) the files from the temporary download folder into the ```resources/i10n``` folder
* some finishing touches to ```fetch``` where is also downloads the translation status of the files among others.
* ```pull``` where I'm thinking of running ```build```, ```fetch``` and ```apply``` in that order.

I see no point in integrating any of these into the normal Maven build process (even though we can if we will), so they will all be run like this:
```
mvn ums-crowdin:<goal>
```

All but ```apply``` will require a valid crowdin API key stored in ```~/.m2/settings.xml```. I've also built in some rather paranoid protection measures into ```push``` since any strings not included in the uploaded file will be deleted (with translations) from crowdin. ```Push``` therefore checks the crowdin project name vs the POM project name vs a plugin defined ```pushProjectName``` variable - which all has to match to do a ```push```. In addition, the variable ```confirm``` must be ```yes```, ```true``` or ```confirm```, like this:
```
mvn ums-crowdin:push -Dconfirm=yes
```

We'll need to figure out where to keep this plugin. If you can create another repository under ```UniversalMediaServer``` that would probably be the best @SubJunk ? We'll also need to store the JAR file somewhere that Maven can download it to avoid having to place it in the local Maven repo manually.

I see that there's different practices when it comes to properties files escaping. To get it right, I should know whether or not to escape: ```=```, ```:```, ```#``` and ```!``` in the strings. I don't know if it's needed or recommended, I've seen files both with and without these characters escaped which seem to work.

Any response on my questions is very welcome :wink:

### CODE REVIEW

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers @SubJunk 
As far as I can tell I'm finished with the [ums-crowdin Maven plugin] (https://github.com/Nadahar/crowdin-maven). I've even written documentation in ```README.md```. Please check if out and see if you think it makes sense, usage of the plugin is described there.

I still think it would be better to have this as a separate repository under the [Universal Media Server organization] (https://github.com/UniversalMediaServer) than under my user. All you need to do @SubJunk is to fork [the original plugin] (https://github.com/glandais/crowdin-maven) to a repository under the [Universal Media Server organization] (https://github.com/UniversalMediaServer) and give me write permissions and I will push my changes there.

I've yet to figure out how to best host the plugin JAR, as the easiest is if it's in an online Maven repository, but Maven central repository have all kind of requirements for license, domain rights and artifact signing that I'm not convinced is worth the trouble to figure out just do make this downloadable.

This PR doesn't contain much yet, but the JAR hosting must be solved before I can set up the POM properly.

**Comment by taconaut** [comment]
@Nadahar I had spent some time figuring out how to deploy to a maven repo and in the end it worked with Sonatype. The main issue I've never been able to solve for automated deploy is to select a subset of artifacts being built (e.g. you don't want to deploy UMS-setup.exe). There are a few notes in the [Deploy](https://github.com/taconaut/pms-mlx/wiki/Developers#deploy) section. You'll also find some changes in the pom.xml (e.g. parent pom).
I had tried this to be able to add a maven dependency onto pms when creating plugins, which would have made the developers life easier.

**Comment by Nadahar** [comment]
@taconaut Thank you for some feedback, I began thinking this PR was invisible to all but me :wink: 

In this case I don't need to deploy a subset since this plugin is a standalone project with it's own POM. I'm still confused about the maven repositories though, and I'm not even sure if Sonatype and "central repo" is two different or the same. Doesn't Sonatype require signing and all that other BS?

**Comment by Nadahar** [comment]
I managed to make my own Maven repository simply by uploading the directory structure containing the plugin from my local repository to [my web server] (http://superbum.org/downloads/maven/com/ums/maven/maven-metadata.xml). I had to rename the two instances of ```maven-metadata-local.xml``` to ```maven-metadata.xml``` and then specify my repository as a ```pluginRepository``` in the UMS POM. I also had to add the plugin group to my local settings. But now Maven download the plugin by itself the first time I run it :smile:

**Comment by taconaut** [comment]
@Nadahar 

>  I began thinking this PR was invisible to all but me

:laughing: Luckily you haven't found the secret switch for that.

>  managed to make my own Maven repository

Cool :) Another possibility could also be to handle it like the dependencies which are marked with `not Mavenized`. These are the ones which require
```
mvn com.savage7.maven.plugins:maven-external-dependency-plugin:resolve-external
mvn com.savage7.maven.plugins:maven-external-dependency-plugin:install-external
```

**Comment by Nadahar** [comment]
> Cool :) Another possibility could also be to handle it like the dependencies which are marked with `not Mavenized`. These are the ones which require
```
mvn com.savage7.maven.plugins:maven-external-dependency-plugin:resolve-external
mvn com.savage7.maven.plugins:maven-external-dependency-plugin:install-external
```
I thought of that, but I think it's quite a hassle that we have to use that at all and would rather like to get rid of that than to add to it. It also requires that the JAR is hosted somewhere accessible from the internet, so it's just as easy to put it in the proper folder structure with the repository xml files making it a working Maven repository where you host it. I'm working on a further idea here now with automatic deployment via FTP, which would create/maintain the repository automatically. All we'd need is FTP access to some folder on some public web server and the rest would be automated if all go as planned.

**Comment by Nadahar** [comment]
After further Maven frustrations I've now managed to automatically make maven create/maintain the repository via sftp in the ```deploy``` phase. That means that this plugin should be pretty much ready to use.

@SubJunk I need some input from you here about the hosting. I have the git repo under my github user and have made the Maven repo on my web server. Although it doesn't bother me to have them there, I don't think it's a good idea to depend on resources that's outside UMS/your control. 

Do you have any issues with creating a separate github repo for this? Also, I think the best thing would be to host the Maven repo on the universalmediaserver.com web host. The files are tiny, so it won't consume any noticable bandwidth. 

Any thoughts?

**Comment by SubJunk** [comment]
@Nadahar I've forked the crowdin repo to the UMS team. I'm happy to host the Maven repo too, I haven't done that before though so instructions or links would be cool

**Comment by Nadahar** [comment]
@SubJunk Great. I'm sorry to say, but I changed my mind about the github repo. Forks in github is more than a normal fork, they are tied to the original repo and will always be compared. I've done extensive changes and I won't be relevant to create a PR towards the original repo since I've replaced much of the logic so that it behaves differently (and probably not how the original author would want). It would be better if you delete the fork and just create a new blank github repo called "ums-crowdin". I will then push into that.

Regarding the Maven repo I'll take that on the forum.

**Comment by Nadahar** [comment]
@SubJunk Did you see the comment above and the forum post?

**Comment by Nadahar** [comment]
Then I've implemented versions management in the ums-crowdin plugin. The updated documentation can be found [here] (https://github.com/Nadahar/ums-crowdin/blob/GitIntegration/README.md).

The plugin will automatically use the local checked out branch when communicating with crowdin. Crowdin has support for [versions management/branches] (https://support.crowdin.com/articles/versions-management/) which I consider to be a great advantage.

Normally if one of us works on a UMS branch and adds new strings for translation, we would have to wait for that branch to be merged to master and then upload the new updated ```messages.properties``` before the translators would ever know the new strings existed. We'd have to give them some time and then pull the updated translations.

With branches, we can push the branch to crowdin as soon as the new string(s) are created. It will exist on a branch on crowdin making the duplicates from master/root to automatically be applied to the branch. Only the new string(s) will need translation. If we pull the translations from master in the mean time, the extra strings won't be included - but they will be if we pull from the branch. Once the branch is merged into master and the new ```messages.properties``` is pushed to crowdin the existing translations from the branch is automatically applied to the root/master translation as well. Once the branch is merged, we can simply delete the branch on crowdin (given that the merged master is pushed to crowdin first) and the translations done while the branch was being developed will remain with the root/master translations.

In short, crowdin will present the sum of all different strings from all brances to the translators, and will only delete a translation when it doesn't exist in any of the branch "base" (English US) files.

I've kept this in a branch in the [ums-crowdin repo] (https://github.com/Nadahar/ums-crowdin) and won't merge and tag it until the move to the UMS github organization is complete.

**Comment by SubJunk** [comment]
I just created the repository under this organization :) I can't always respond quickly because I'm very busy lately, I appreciate what you're doing

**Comment by Nadahar** [comment]
@SubJunk Thanks :+1:  

I know that you're busy, I just never know if you've missed it or haven't had the time to reply. Also, I gave my reply to you within 2 minutes of what you wrote and you continued to comment on other issues here after that, indicating to me that you probably hadn't seen it.

I'll get right at "moving in" there :smile:

**Comment by Nadahar** [comment]
Then the ums-crowding plugin is moved into the new github repo: https://github.com/UniversalMediaServer/ums-crowdin.
I've also tagged v0.1 and v0.2, and as soon as we get our Maven repo up and running this should be good to go.

**Comment by Nadahar** [comment]
I'm currently working removing reliance on default encoding in UMS (conversion between bytes and char/String that doesn't specify what encoding to use and that could fail on non US OS locales). This is tricky work since I don't always know exactly what kind of data I'm dealing with, and could introduce bugs. I'll note any questions that arise her as they come up:

* I'm assuming that the base64 encoding used adheres to ASCII codes [here] (https://github.com/UniversalMediaServer/UniversalMediaServer/blob/master/src/main/java/net/pms/dlna/LibMediaInfoParser.java#L736). As I understand that's usually the case with base64, but since MediaInfo encodes this it's more than I can take to find out for sure.

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers @SubJunk 
I really need some input on this one. I'm looking at the encoding of the AviSynth script created by UMS in ```AviSynthFFmpeg.getAVSScript()```. First of all I don't really know how AviSynth scripts are used or what purpose they server, but I understand as much that they don't support unicode, but that if UTF-8 encoded without BOM they will work sometimes - depending on what's being called? [Here] (http://stackoverflow.com/questions/25364394/utf-8-source-files-are-not-supported-in-avisynth) is an example using ```ffmpegsource2``` whatever that is. We're using it for FFmpeg here, so it seems partly relevant.

If we don't UTF-8 encode these files, file names with "non english characters" will fail. At the same time, it's no use to UTF-8 encode if whatever reads the files doesn't interpret UTF-8.

In addition I saw [this line] (https://github.com/UniversalMediaServer/UniversalMediaServer/blob/master/src/main/java/net/pms/encoders/AviSynthFFmpeg.java#L105). This is clearly flawed for anything but Windows. Although it's easy to fix, how come that this still exists after so many years? Isn'n this code in use anymore?

Should I fix the line, and what do I do with the encoding?

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers @SubJunk 
Then I've managed to release [ums-crowdin] (https://github.com/UniversalMediaServer/ums-crowdin) to Maven central as well. That means that the plugin is available for download, and I can continue this PR.

**Comment by SubJunk** [comment]
Brilliant!

**Comment by Nadahar** [comment]
@SubJunk I see that you had turned off "Build pull requests" at Travis do avoid the double build. I'm not sure it's such a good idea, and I've turned it back on (I didn't think I had the permissions to do so, so I'm still not sure if I really have, but it says so).

The way I understand it is that we can choose to build both pushes to the repository (that is, each time one of us pushes a commit to any branch, including master) and pull requests. We'd want to build pushes since that can will apply to commits made directly at master among others, and commits made in other branches. We'd also want to build pull requests coming from forks. In both these cases only one build will be done. 

But, when we make a branch at this repo, create a pull request for that branch and then push to that branch, it will trigger both and do a "double build". Even though it's unnecessary, there's no way to avoid it without either not building "external" pull requests or our own pushes. Except from the wasted resources at Travis from doing it twice, it has no consequences. The builds will be identical and both should either fail or succeed. Ideally, this is something Travis should avoid for their sake, not something that should worry us.

Actually it has helped me a couple of times when testing the "external dependencies" PR. In some cases one of the builds would fail and not the other. In those cases I could always rebuild the failed one and it would succeed. When investigating it would be due to some download to fail during build, some external server not aswering in time or similar. Without this double check, I would have scratched my head a bit extra when trying to figure out why it failed when there really was nothing wrong with the build. As a consequence we should remember that if a build failure doesn't have an obvious reason, we should try to restart it at Travis. Sometimes they fail for external reasons and our code could be fine.

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers @SubJunk 
As mentioned further up in the PR I wanted to remove all reliance on default encoding in UMS as a part of this PR. That means that any ```byte``` <-> ```String``` conversion should have a explicitly defined encoding (UTF-8, ASCII, ISO 8859:1 etc.). 

The way it's solved today is that the JVM default encoding is set to UTF-8 in the wrapper for Windows, or in scripts for OS X/Linux if I understand it correctly. The code then omits to specify encoding when doing conversion and relies on the JVM default encoding. 

I think this is bad for a couple of reasons. First of all, not all encoding SHOULD be UTF-8 - it depends on the source or destination for the data. Neither do I like relying on the encoding argument being there when UMS is launched, people could launch UMS in other ways leading to bugs that could be a headache for us to figure out. But, most importantly we loose the information that the default encoding SHOULD have, the actual encoding used on the OS. We need this information to correctly parse OS output, but at the moment this information isn't available to us because we overwrite it when we launch UMS. On top of this, FindBugs warns about all reliance on default encoding making it harder to use this to find "real" bugs.

But, I've realized that this is a much bigger task than I originally thought. What I've decided to do is therefore to include the few fixes to reliance on default encoding that I've already made in this PR, but leave the rest for later. It's more important to get this PR and the crowdin integration finished, and the default encoding reliance removal will have to be it's own future task.

That means that I'll stop all work on this now, and start implementing the language choice functionality. When that's done, this PR is good to go as far as I know.

**Comment by Nadahar** [comment]
@SubJunk The Travis team replied. It seems there are multiple issues on their issuetracker around the subject, from which I think https://github.com/travis-ci/travis-ci/issues/1147 and https://github.com/travis-ci/travis-ci/issues/3276 are the best matches. I don't really understand why, but the Travis team seem to think it's necessary. If it doesn't bother them I guess it shouldn't bother us?

**Comment by Nadahar** [comment]
Then I've made the logic for building the list of languages used for language selection, here is an attempt at describing the logic:

The sorting places the default/recommended choice on top of the list, and then tried to place other relevant choices close to the top in descending order by relevance. The rest of the list is alphabetical by the preferred/currently selected language's language names.
The sorting is done following these rules:
* The base language (en-US) and the language closest matching ```preferredLocale``` is looked up. If the closest matching language has a coverage greater or equal to ```defaultTranslatePct``` or approval greater or equal to ```defaultApprovedPct``` it will be placed on top. If not, the base language will be placed on top. Whichever of these is not placed on top is placed second. If a closely matching language cannot be found, only the base language will be placed on top.
* A search for related languages is performed. Related is defined by either having the same language code (e.g ```en```) or the same country code as ```preferredLocale```. Related languages are then sorted descending by coverage and put after that or those language(s) placed on top.
* The rest of the languages are listed alphabetically based on their localized (based on ```preferredLocale``` names.

If the localized language name differs from the English language name, the English language name is shown in parenthesis. This is to help in case the localized names are incomprehensible to the user.

In addition to that, translation coverage is shown in percent in parenthesis for any non-English languages. Since English language variants only override the strings that differ from US English, coverage is impossible to calculate for them.

I've also added support for ```SuppressFBWarnings``` annotations to be able to suppress them where they give false positives. This doesn't seem to work with ```findbugs-maven-plugin```, but I've [opened an issue about it] (https://github.com/gleclaire/findbugs-maven-plugin/issues/34).

**Comment by Nadahar** [comment]
I've been struggling with the sizing of the "select language" dialog for some time now. I've finally figured out [what goes wrong] (http://bugs.java.com/view_bug.do?bug_id=4924163) - but as usual with Java no fix or workaround is offered. This bug is from 2004....

I think I'll have to try a different component than ```JTextArea```.

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers @SubJunk 
I finally got through fighting with Swing. The latest changes are relatively well explained in the commit messages, so I won't repeat it here.

In short, what I've done is to create a dialog that pops up before the configuration wizard, the splash screen and any other dialog other than then profile selection (as it needs a loaded configuration) if the language isn't set in the configuration.

The dialog puts the "best suggestion" on top, based on the OS locale and the translation coverage for that language (suggesting US English if coverage is below a certain threshold). Coverage is not calculated for English languages, they are always considered 100%. This is because of the way we handle English, only overriding those strings that differ from the base language, US English.

If any language that is below the coverage threshold is selected, a warning will appear informing of the incomplete translation and that missing strings will be shown in US English.

Any language that has a coverage below a minimum threshold will not show up in the list at all. The only way to use these languages is to configure UMS.conf manually and set them there, they can not be selected from the GUI. The idea is that when their translation coverage is low enough, they aren't really good for anything. At the same time it allows us to add languages without worrying about users picking those before they hold a certain standard. The minimum threshold is currently only 20%, a limit I find way too low. But, many of the current UMS languages are in the twenties, so setting it higher would exclude them. We should raise it if you feel so, I'd say 50% would be a more reasonable minimum threshold.

Translation coverage is based on the number of translated strings, not words. 

The dialog also shows a little text in the bottom informing about the crowdin project and that anyone can contribute, with a hyperlink.

The dialog is dual languaged - all texts are shown in the selected language and in English. This is to ensure that selecting some completely unreadable language doesn't preventing selecting something different.

I've also replaced the language selection in the general tab with this dialog, and added a link to the crowdin project on the about tab.

Not counting any bugs I've missed this is ready for testing and finally merging. I'm planning the look through the new code and see if I can create any tests for any of it, but that shouldn't prevent testing/reviews/feedback.

**Comment by Nadahar** [comment]
I'd also like feedback/corrections on the English strings I've added to ```messages.properties``` to ensure it's proper English since I'm not native.

**Comment by SubJunk** [comment]
@Nadahar I'm testing it now. Small point, but it would be cool if the window popped up in the middle of the screen. I think our other windows do that but this one appeared off to the right

**Comment by Nadahar** [comment]
@SubJunk It's centered on my screen. Do you have a dual monitor setup? Actually, I don't place it - I use ```JOptionPane``` - the same class that shows most dialog boxes including the splash screen, and I've got no idea how to control the position. Maybe it's centering on the UMS Frame?

**Comment by SubJunk** [comment]
When I clicked Select, the UMS window showed up centered as usual so I don't see why it would be like that

capture

**Comment by SubJunk** [comment]
I suggest changing and shortening the text to something like "If you would like to help us improve the language translations, please visit our Crowdin page to contribute easily.", where "our Crowdin page" links to our Crowdin page.

**Comment by Nadahar** [comment]
@SubJunk I'll have to look at the centering tomorrow, could it be that I didn't notice this? I see a possible cause because the dialog is dynamicly sized. Maybe the centering is done at one stage and then the box grows when the texts are added, making it appear east and south of the center point. I don't know, this smells like another deep dive into Swing, something I'm planning to quit doing ;)

About the text, is the change because my English was bad or because you didn't like it's content?

**Comment by SubJunk** [comment]
@Nadahar if you want I can have a look at the centering, and your English is great! I'm not sure I like the idea of having a reminder that UMS is made by volunteers on the first page - it's true but it seems like a disclaimer like "hey if the product is bad you can't blame us because we're just volunteers" :) I know it's not meant that way but I worry it will be interpreted that way. I also don't think the text needs to be so wordy

**Comment by Nadahar** [comment]
@SubJunk I've solved the centering, I just haven't pushed it yet as I wanted to do the sentencing in the same go.

Ok, I suspected that wording didn't come across as I wanted. In Norwegian we have a word called "dugnad" and I haven't quite found a corresponding English term. This is what I tried to express. It's like collaborative effort but also implies that you work for free. It's seen as a positive thing, something people do for their friends and some places neighbours, where they all "pull together". My idea was to express that UMS is made as a sum of free contributions from people, and that he or she who reads that are welcome to do their little part.

Is that a bad idea? If not, any idea how to express it without writing a small book?

**Comment by SubJunk** [comment]
Ah I see, that's a cool word. We should have one like that. How about:
"These translations are contributed to us via the Crowdin community. If you would like to help translate, please visit our Crowdin page."

**Comment by Nadahar** [comment]
```Universal Media Server is made entirely by volunteers. If your language is missing or incomplete, you can contribute with translations. It's easy, you can do it from your web browser. Simply go to the <a href="%s">crowdin.com project page</a> and start translating.```

I just added my committed text for reference, so I don't have to look elsewhere to see what's not there :wink: 

It seems to me that you wish to get a slightly different message across than I do. My idea was to say that anyone can contribute, and that UMS is made of contributions. It might just be me, but I think giving them some "ownership" in it in the way that if I gave my contribution, I'm a part of it too. To me that's a better incentive than that "you can choose to help us". I also wanted to express that languages that are not in the list can still be contributed to. Partly because with this new setup it's easy to integrate new languages, and partly because existing languages with a low coverage will be hidden.

The reason I'm writing ```crowdin``` instead of ```Crowdin``` is because as far as I can tell that's the way they use it. Some companies think that is hip I guess, it's not the first one I see using just small letters, and I precieve it as part of the "trade mark".

I just feel that your take is a bit more "us" and "them" than what I'm trying to say, but I guess that's up to you in the end. I'm not so worried as being percieved as "professional" as I am inclusive.

**Comment by Nadahar** [comment]
Another thing I'm unsure about is the layout on the General page. It's ugly, but so is the whole layout, but I still feel that this made it worse. 

At first I combined ```Language:``` and the name of the language into one label, so it would say e.g ```Language: Norwegian``` (but in Norwegian of course). I then had the ```select language``` button in the same column as the ```server name``` is.

But then I figured out that logically this was wrong, as the data seemed to be presented in the second column and modifications/options in the third. Now I regret it, as I think it looks worse even though it slightly better conforms to the logic of the layout. I'm not sure that the logic is that clear anyway, and that trying to stick with that is worth the even messier impression. Should I revert it?

**Comment by Nadahar** [comment]
@SubJunk 

What about "If your language is missing or incomplete, you can contribute with translations. It's easy, you can do it from your web browser. Simply visit [our crowdin page] and start translating."?

**Comment by SubJunk** [comment]
That looks good, a few minor changes would make it:

If your language is missing or incomplete, you can contribute translations. It's easy and you can do it from your web browser. Simply visit [our crowdin page] and start translating.

**Comment by Nadahar** [comment]
> If your language is missing or incomplete, you can contribute translations. It's easy and you can do it from your web browser. Simply visit [our crowdin page] and start translating.

Done deal :+1: 

@SubJunk What do you thing about the layout on General tab that I mentioned above?

**Comment by SubJunk** [comment]
I agree it doesn't quite look right. Maybe something like this would be more consistent with our interface

language

The dropdown was even better but I guess we want to display the whole popup instead of just the dropdown?

**Comment by Nadahar** [comment]
@SubJunk That was the idea yes, there's so much logic built into the dialog that replicating that in the General tab would mess things up completely. The dialog also offers to restart UMS when exited if the language was changed. The dropdown is still there, it's just moved to the dialog. I think your suggestion looks good, I'll try to mimic it.

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers @SubJunk 
I've added a few tests (most of this is GUI, and that's really hard to test) and implemented the changes discussed above.

Further testing/views are appreciated.

**Comment by Nadahar** [comment]
@SubJunk Could you test the updated version and see if the issues you've seen are resolved in a satisfactory way?

I've force pushed the branch, so you should probably use ```git reset --hard <reponame>/crowdin``` to update your local branch.

**Comment by SubJunk** [comment]
I just tested it and it seems great, I'm happy for it to be merged if the others are. Nice work

**Comment by valib** [comment]
@SubJunk no objection from me.

**Comment by Nadahar** [comment]
I've now tested it on Ubuntu, Centos 7 and OS X 10.8. I discovered a minor thing on OS X (the text boxes had a slightly different background color than the panels for some reason) which I've fixed. I've then retested on Windows 7, Ubuntu 12, Centos 7 and OS X 10.8.

If noone has any objections by 14:00 zulu time, I'll merge.

**Comment by Nadahar** [comment]
@UniversalMediaServer/developers I've made instructions in the private forum as exactly how to use the crowdin plugin in our case. General instructions can be found [here] (https://github.com/UniversalMediaServer/ums-crowdin/blob/master/README.md).

**Comment by Nadahar** [comment]
Ok, here we go - merging.

**Comment by SubJunk** [comment]
:D

**Comment by skeptical** [comment]
> Further testing/views are appreciated.

@Nadahar late to this party but my 50 cents: I know you have (admirably) moved heaven and earth building the tower/dialog of babel here, but I honestly feel it would be cleaner to revive/improve the general tab language selection combobox with a 'Help translate UMS' button next to it (to open up a simple dialog containing the 'About translations' info), i.e. ```-=+``` (less is more).

Motivated translators will find what they need, and ordinary users will get a quick ordinary control similar to other settings.

Btw translators on the whole don't need a lot of encouragement (as you know better than me, e.g. crowdin response), since seeing poor language is highly irritating.  And highly subjective, when I was googling for Farsi translation resources I came upon the Kodi farsi discussion thread which turned out to be largely people calling each other idiots one way or another :).

**Comment by Nadahar** [comment]
@skeptical The dialog is just an ordinary dialog, the fact that I struggled a lot with Swing making it doesn't make it special in any way. I read someone say somewhere:
> Swing makes hard things possible and simple things hard.

I guess there's some truth to that.

In any case, I didn't make that dialog just to recruit translators. In fact, that was an "afterthought". The major point with the dialog is that it pops up during UMS startup when no language is set, instead of the previous behaviour of just selecting whatever language your OS locale dictates. I understand that the previous behaviour work perfectly for anyone with English OS locale, but for anyone else it was a disaster. It just makes UMS seem like an application that's 20% finished, and to be honest that alone almost make me skip UMS alltogether when I first installed it. What that says to me is that anyone who's willing to let this be the default, is so detached from any kind of quality standard that the rest of the software is bound to be a nightmare.

It gives a recommended language based on translation statistics gathered from crowdin, and it warns the user if a language with too low translation coverage is selected. This was my main motivation for making this, to try to ensure that the users get a better experience when using UMS. The dialog is also "multi language" in the sense that texts are displayed both in the selected language and in English to avoid the typical "trap" where you select a to you completely unreadable language and is unable to get it back because it's impossible to navigate the "garbled" interface. 

These things, the multiple languages and the warning needs space, and that was my main reason for choosing to replace the combobox with the dialog. My plan wasn't to do that at all, my plan was to use the dialog during startup and then make corresponding functionality on General tab.

But, fighting Swing every step made it drag out into eternity, and I felt the need to finish this and get crowdin integrated. That, and the fact that I quickly realized when I started to look at the layout in General tab that there simply was no space there for this without drastically reorganizing the whole layout made me come to the conclusion that it was better to just call the dialog from there.

I honestly can't see the big problem, I doubt most users will change language that often that the two extra clicks will hurt them.

When it comes to recruiting translators I disagree. Yes, there's been some response and we've gotten quite a lot of new translations. That said, the overall coverage is still really bad as you can see [here] (https://crowdin.com/project/universalmediaserver).

I simply don't think there's enough "motivated translators" using UMS, and those that are need to find out how easy it is to contribute without doing their own investigation. I doubt most users knows anything about "the open source mentality" and contributing, they probably think it's all so complicated that they don't even consider it. That's what I'm trying to change with the "About translations" paragraph.

**Comment by skeptical** [comment]
@Nadahar Firstly happy new year :) and secondly I understand the chain of events/reasoning etc, it's very natural, and thirdly it's your baby so I defer.  My only point is that (taking it a bit further) nothing prevents the combobox from appearing both in the startup dialog (with explanations, etc) and also on the general tab (with 'Help translate UMS' button alongside), this would be a more clean/natural interface imo, and probably attract translators just as well.

Side point: crowdin statistics are a bit deceptive (counting words I guess?), e.g.  it considers fa to be 42% translated, although ```467 items (messages_fa.properties) / 567 items (messages.properties) = .82``` which is a more valid metric imo.

**Comment by Nadahar** [comment]
Happy new year. I thought I'd picked one that counted phrases, but it seems not. Here's the full collected statistics for Farsi:
```
fa.name=Persian
fa.phrases=578
fa.phrases.approved=432
fa.phrases.translated=466
fa.progress.approved=41
fa.progress.translated=42
fa.words=3597
fa.words.approved=1491
fa.words.translated=1531
```
Changing which of these numbers are used is quick, both are available. Still, since the difference is that big, it seems that most of the long sentences have been skipped...

**Comment by skeptical** [comment]
@Nadahar yep, I confess I skipped all the wordy tooltips ('tip' != 'essay' imo :-).
